tensorzero-core/src/tool.rs

```diff
@@ -410,67 +439,80 @@ impl ToolCallConfig {
             dynamic_parallel_tool_calls,
             dynamic_provider_tools,
         } = args;
-        // If `allowed_tools` is not provided, use the function's configured tools.
+        // If `allowed_tools` is not provided, use the function's configured tools plus any dynamic tools.
         // This means we allow all tools for the function.
-        let mut allowed_tools = match dynamic_allowed_tools {
+        let allowed_tools = match dynamic_allowed_tools {
             Some(allowed_tools) => AllowedTools {
                 tools: allowed_tools,
-                choice: AllowedToolsChoice::DynamicAllowedTools,
```

Comment:
IIUC the behavior change is:
- Before, with `allowed_tools`, the model is allowed to pick any of {function tools, allowed tools} (after the change this is specified in `dynamic_additional_tools`)
- After, with `allowed_tools`, the model is only allowed to pick from {allowed tools}

We should warn loudly in the docs that this is breaking.